include/circt/Dialect/Arc/Ops.td

```diff
@@ -448,4 +452,20 @@ def LutOp : ArcOp<"lut", [
   let hasVerifier = 1;
 }
 
+def ZeroCountPredicateLeading  : I32EnumAttrCase<"leading", 0>;
+def ZeroCountPredicateTrailing  : I32EnumAttrCase<"trailing", 1>;
+def ZeroCountPredicate : I32EnumAttr<
+    "ZeroCountPredicate", "arc.zero_count predicate",
+    [ZeroCountPredicateLeading, ZeroCountPredicateTrailing]> {
+  let cppNamespace = "circt::arc";
+}
+
+def ZeroCountOp : ArcOp<"zero_count", [Pure, SameOperandsAndResultType]> {
```

Comment:
LLVM is all over the place here.  CLZ and CTZ I think are separate, but all the comparisons are enumerated this way.  Just curious why you went with the enum.

Reply:
@maerhart

Reply:
1. CLZ and CTZ are very similar and having them in one op could help code sharing. E.g. canonicalization: one concrete example are the priority encoders which are a sequence of `comb.extract` and `comb.mux`, the extractions can go from high to low or the other way around. When this is not easy to detect immediately (e.g. when there is a function boundary between) we possibly have to reverse the order afterwards (as a canonicalization), we don't care about what the initial value of the enum is, we just want to flip it (this also saves us from building a new op and changing the use/def chain of the ssa values in that case).
2. I expect that most of the patterns/passes where we want to match on one of these, we want to match on both. That means, if we have separate ops we would need to do `isa<CLZ, CTZ>` or two separate dynamic casts, or define two separate patterns (possibly using templating) when using the dialect conversion/greedy rewrite framework or one that matches on generic `Operation`s.
3. That way it seems more consistend w.r.t. `comb.icmp` to me.

But I'd also be interested in your thoughts on this @darthscsi